runtime_scan/pkg/provider/gcp/common.go

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/googleapis/gax-go/v2/apierror"
+	"google.golang.org/api/googleapi"
+
+	"github.com/openclarity/vmclarity/runtime_scan/pkg/provider"
+)
+
+func handleGcpRequestError(err error, actionTmpl string, parts ...interface{}) (bool, error) {
+	action := fmt.Sprintf(actionTmpl, parts...)
+
+	if e, ok := err.(*apierror.APIError); ok {
+		origError := e.Unwrap()
+		if e, ok := origError.(*googleapi.Error); ok {
+			sc := e.Code
+			switch {
+			case sc >= 400 && sc < 500:
+				// Client errors (BadRequest/Unauthorized etc) are Fatal. We
+				// also return true to indicate we have NotFound which is a
+				// special case in a lot of processing.
+				return sc == http.StatusNotFound, provider.FatalErrorf("error from gcp while %s: %w", action, err)
+			default:
+				// Everything else is a normal error which can be
+				// logged as a failure and then the reconciler will try
+				// again on the next loop.
+				return false, fmt.Errorf("error from gcp while %s: %w", action, err)
+			}
+		} else {
+			// Error should be a googleapi.Error.
+			return false, provider.FatalErrorf("unexpected error from gcp while %s: %w", action, err)
+		}
+	} else {
+		// Error should be a apierror.APIError.
+		return false, provider.FatalErrorf("unexpected error from gcp while %s: %w", action, err)
+	}
+}
+
+func ensureDeleted(resourceType string, getFunc func() error, deleteFunc func() error, estimateTime time.Duration) error {
+	err := getFunc()
+	if err != nil {
+		notFound, err := handleGcpRequestError(err, "getting %s", resourceType)
+		// NotFound means that the resource has been deleted
+		// successfully, all other errors are raised.
+		if notFound {
+			return nil
+		}
+		return err
+	}
+
+	err = deleteFunc()
+	if err != nil {
+		_, err := handleGcpRequestError(err, "deleting %s", resourceType)
+		return err
+	}
+
+	return provider.RetryableErrorf(estimateTime, "%s delete issued", resourceType)
+}
+
+// example: https://www.googleapis.com/compute/v1/projects/gcp-etigcp-nprd-12855/zones/us-central1-c/machineTypes/e2-medium -> returns e2-medium
+func getLastURLPart(str *string) string {
+	if str == nil {
+		return ""
+	}
+
+	spl := strings.Split(*str, "/")
+
+	if len(spl) == 0 {
+		return ""
+	}
+
+	return spl[len(spl)-1]
```

Comment:
maybe better to use `net/url` library?

https://go.dev/play/p/K_er3w-z5fr

https://stackoverflow.com/questions/59438561/how-to-obtain-the-last-path-segment-of-a-url